### PR TITLE
Deployment fail error podevents [WIP]

### LINKED ISSUE #6284: Readiness check failure detection
## Overview of feature request

When the readiness check for a pod fails, no further traffic is sent to the pod. However, the pod does not restart. Due to this, it is hard to determine the readiness check failure as a terminal condition.

Once #6283 is implemented, instead of simply reporting "deployment timed out", we want to watch pod events and report possible causes of failure. For example, readiness check failure is reported in the pod events and could be a possible cause for deployment timeouts which might give the user an idea how to troubleshoot the root cause. 

## Additional context

[AB#9445](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/9445)

### PR DESCRIPTION
**Pull request by vinayada1**
# Description

Readiness probe failure is a non-terminal failure. As a result, currently, we wait for the deployment to complete or eventually timeout. In case of a timeout, the user has no clue why the deployment timed out. The purpose of this change is to detect the readiness probe failure and provide this information to the user if the deployment times out.

## Type of change

- This pull request adds or changes features of Radius and has an approved issue (https://github.com/radius-project/radius/issues/6284).

Fixes: #6284 #6835 

### Summary
🚀🐛🧪

This pull request adds support for reporting the status and possible failure causes of asynchronous operations when deploying Kubernetes resources. It modifies the `AsyncOperationRequest`, `AsyncOperationStatus`, `DeploymentProcessor`, and `resource_handler` types and functions to pass and update the operation status. It also updates the `kubernetes_deployment_waiter` and `kubernetes_http_proxy_waiter` modules to use the pod events API to check the readiness probe status of the deployments. It adds and modifies some unit tests and functional tests to cover the new features. It also updates the `rbac.yaml` file and the specification file for the gateways resource.

> _We're deploying to the Kubernetes cluster_
> _With async operations and readiness probes_
> _We'll update the status and handle the errors_
> _Heave away, me lads and lasses, heave away_

### Walkthrough
*  Add `events` resource to `rp` role in `rbac.yaml` to allow `rp` service account to access pod events ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-13e38bac99402c4ae82609aaba31139f5f9f7c229d1bd955032255cb18f356c6R18))
*  Reduce default async operation timeout from 120 to 20 seconds in `request.go` to make test cases run faster and fail earlier ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-2a4ba42fee4a75a9e39ee8770bd4bcafde486b06d23c1002f335794bc2a62645L30-R30))
*  Add `OperationStatus` field to `AsyncOperationRequest` struct in `request.go` to hold the current status of the async operation ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-2a4ba42fee4a75a9e39ee8770bd4bcafde486b06d23c1002f335794bc2a62645R59-R60))
*  Assign `OperationStatus` field of `AsyncOperationRequest` to `OperationStatus` field of `AsyncOperationStatus` in `statusmanager.go` to pass the operation status from the request to the status manager ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-aac87dc11a514274bedf43d8d447279fccca2f833544b38124ffe2d45519a3bbR219))
*  Add `OperationInfoKey` constant to `worker.go` to use as a key to store and retrieve the operation information from the context ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-5d9b9964242d77606f4868e3fe049978ea1c3f1903afcfeb60c76e19d9785fc5R62-R63))
*  Initialize operation information to an empty string in the context before running the operation in `worker.go` ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-5d9b9964242d77606f4868e3fe049978ea1c3f1903afcfeb60c76e19d9785fc5R173))
*  Initialize `OperationStatus` field of `AsyncOperationRequest` to an empty `AsyncOperationStatus` struct in `worker.go` to avoid nil pointer dereference ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-5d9b9964242d77606f4868e3fe049978ea1c3f1903afcfeb60c76e19d9785fc5R231-R232))
*  Print the result of the operation to the standard output for debugging purposes in `worker.go` ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-5d9b9964242d77606f4868e3fe049978ea1c3f1903afcfeb60c76e19d9785fc5R260))
*  Remove an empty line from `runOperation` function in `worker.go` to improve code readability and style consistency ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-5d9b9964242d77606f4868e3fe049978ea1c3f1903afcfeb60c76e19d9785fc5L272))
*  Append the operation status to the error message when the operation times out and print a message when the cancel function is invoked in `worker.go` ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-5d9b9964242d77606f4868e3fe049978ea1c3f1903afcfeb60c76e19d9785fc5L288-R299))
*  Add `operationStatus` parameter to `Deploy` function signature in `deploymentprocessor.go` to allow the deployment processor to receive and update the operation status ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-2f71b55e0dceaf541f852a86c0b54060819c6266f5257474fa863826d4043815L57-R57), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-2f71b55e0dceaf541f852a86c0b54060819c6266f5257474fa863826d4043815L247-R250))
*  Print the operation status to the standard output for debugging purposes in `deploymentprocessor.go` ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-2f71b55e0dceaf541f852a86c0b54060819c6266f5257474fa863826d4043815R163), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-2f71b55e0dceaf541f852a86c0b54060819c6266f5257474fa863826d4043815L247-R250))
*  Pass the operation status from the caller to the `deployOutputResource` function in `deploymentprocessor.go` ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-2f71b55e0dceaf541f852a86c0b54060819c6266f5257474fa863826d4043815L288-R290))
*  Start the events informer in `kubernetes_deployment_waiter_test.go` to get the pod events from the Kubernetes API in the test cases ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-f21cc2a2965eb59afa1905cb56125a769b00a476a0cb1a785c298dc6c9c57b58R102))
*  Skip the test cases that are not relevant for the readiness probe scenario in `kubernetes_deployment_waiter_test.go` ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-f21cc2a2965eb59afa1905cb56125a769b00a476a0cb1a785c298dc6c9c57b58L438-R539), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-f21cc2a2965eb59afa1905cb56125a769b00a476a0cb1a785c298dc6c9c57b58L552-R581))
*  Modify the type of the `doneCh` variable from `chan error` to `chan deploymentStatus` and the logic of receiving the error from the `doneCh` channel in `kubernetes_deployment_waiter_test.go` to use a custom struct `deploymentStatus` that contains both the possible failure cause and the error value ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-f21cc2a2965eb59afa1905cb56125a769b00a476a0cb1a785c298dc6c9c57b58L630-R633), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-f21cc2a2965eb59afa1905cb56125a769b00a476a0cb1a785c298dc6c9c57b58L697-R700), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-f21cc2a2965eb59afa1905cb56125a769b00a476a0cb1a785c298dc6c9c57b58L774-R777), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-f21cc2a2965eb59afa1905cb56125a769b00a476a0cb1a785c298dc6c9c57b58L848-R851), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-f21cc2a2965eb59afa1905cb56125a769b00a476a0cb1a785c298dc6c9c57b58L927-R930), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-f21cc2a2965eb59afa1905cb56125a769b00a476a0cb1a785c298dc6c9c57b58L1009-R1012), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-f21cc2a2965eb59afa1905cb56125a769b00a476a0cb1a785c298dc6c9c57b58L1098-R1101))
*  Modify the logic of calling and checking the error from the `checkDeploymentStatus` function in `kubernetes_deployment_waiter_test.go` to call the function synchronously and use the return value directly ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-f21cc2a2965eb59afa1905cb56125a769b00a476a0cb1a785c298dc6c9c57b58L783-R789), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-f21cc2a2965eb59afa1905cb56125a769b00a476a0cb1a785c298dc6c9c57b58L934-R942))
*  Add a block of code to create a pod event with the message "Readiness probe failed" in `kubernetes_deployment_waiter_test.go` and comment it out for future use ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-f21cc2a2965eb59afa1905cb56125a769b00a476a0cb1a785c298dc6c9c57b58R1119-R1134))
*  Add `operationStatus` parameter to `addEventHandler` and `waitUntilReady` function signatures in `kubernetes_deployment_waiter.go` to receive and update the operation status when waiting for the deployment to be ready ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L41-R53), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L62-R77))
*  Modify the type of the `doneCh` variable from `chan error` to `chan deploymentStatus` and the logic of receiving the error from the `doneCh` channel in `kubernetes_deployment_waiter.go` to use a custom struct `deploymentStatus` that contains both the possible failure cause and the error value ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L62-R77))
*  Add a variable `possibleFailureCauses` to store a slice of strings that represent the possible failure causes of the deployment and update it by receiving the `deploymentStatus` values from the `doneCh` channel in `kubernetes_deployment_waiter.go` ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L62-R77))
*  Append the operation status to the error message when the context is done and set the operation information to the context using the `OperationInfoKey` constant and the `possibleFailureCauses` slice in `kubernetes_deployment_waiter.go` ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L62-R77))
*  Pass the operation status from the caller to the `startInformers` and `checkDeploymentStatus` functions in `kubernetes_deployment_waiter.go` ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L73-R83), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L79-R144))
*  Add `operationStatus` parameter to `checkDeploymentStatus` function signature in `kubernetes_deployment_waiter.go` to receive and update the operation status when checking the deployment status ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L120-R160))
*  Modify the type of the `doneCh` parameter from `chan<- error` to `chan<- deploymentStatus` and the logic of sending a nil error or an error to the `doneCh` channel in `kubernetes_deployment_waiter.go` to use a custom struct `deploymentStatus` that contains both the possible failure cause and the error value ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L120-R160), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L138-R181), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L147-R205))
*  Add `operationStatus` parameter to `checkAllPodsReady` function signature in `kubernetes_deployment_waiter.go` to receive and update the operation status when checking the pod readiness ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L217-R263))
*  Modify the type of the `doneCh` parameter from `chan<- error` to `chan<- deploymentStatus` and the logic of sending a nil error or an error to the `doneCh` channel in `kubernetes_deployment_waiter.go` to use a custom struct `deploymentStatus` that contains both the possible failure cause and the error value ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L217-R263), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L229-R281))
*  Add `informerFactory` parameter to `checkPodStatus` function signature in `kubernetes_deployment_waiter.go` to use the informer factory to get the pod events from the Kubernetes API ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L269-R316))
*  Modify the type of the return values from `(bool, error)` to `(bool, *deploymentStatus)` and the logic of returning a nil error or an error in `kubernetes_deployment_waiter.go` to use a custom struct `deploymentStatus` that contains both the possible failure cause and the error value ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L269-R316), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453R339), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L298-R349), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L305-R355))
*  Add a block of code that checks the pod events when the container state is `Waiting` with a non-terminally failed reason or `Running` but not `Ready` or when the pod condition is not `PodReady` in `kubernetes_deployment_waiter.go` and return a `deploymentStatus` value with a possible failure cause and a nil error if the pod event contains the message "Readiness probe failed" ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L269-R316), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-864c8230511d8d335b1939ed9fa8196ba64cb535aac443d63474d3c309024453L311-R420))
*  Add `operationStatus` parameter to `addEventHandler` and `waitUntilReady` function signatures in `kubernetes_http_proxy_waiter.go` to receive and update the operation status when waiting for the http proxy to be ready ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-397d63d65f44b2c85cd035e75632e13cbcd0d89c0610f44e818bd5bd809e1138L64-R68))
*  Add `OperationStatus` field to `PutOptions` struct in `resource_handler.go` to hold the current status of the async operation ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-8b0006aeb2f47650c6499d4337299a9ff29a8ff849c55d68ddc9f6d290e680f6R51-R53))
*  Add `operationStatus` parameter to `waitUntilReady` function calls in `kubernetes.go` to pass the operation status from the caller to the deployment waiter or the http proxy waiter ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-033dc8a7c9b970cc30f7420ce8cf2f12a26258b276779683ba3f761c481bf0a7L127-R130), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-033dc8a7c9b970cc30f7420ce8cf2f12a26258b276779683ba3f761c481bf0a7L134-R136))
*  Update the path and the version of the specification file for the gateways resource in `loader_test.go` to use the latest version that supports the readiness probe feature ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-7b8cf0fed92b04574159896aa3e609b874fb6822eb4943c56e1f07966c46c089L59-R64))
*  Add a test function `Test_Container_FailDueToBadReadinessProbe` and a template file `corerp-resources-container-bad-readinessprobe.bicep` to test the scenario where the container resource deployment fails due to a bad readiness probe ([link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-9442e1a784d8aba972917e5cfd69a85102d0b1393b88dc2be87e84a47c794970R357-R393), [link](https://github.com/radius-project/radius/pull/6862/files?diff=unified&w=0#diff-b87aaf814e31b3a182cf66d3c34ff2639779cb5c0d9019cd127ad3d8ea27dd5cR1-R53))

### CODE REVIEW

**Comment by rynowak** [comment]
Hi @vinayada1 - what's the plan for this PR? There's no description and no associated work-item.

**Comment by vinayada1** [comment]
> Hi @vinayada1 - what's the plan for this PR? There's no description and no associated work-item.

Added this.

File: pkg/corerp/backend/deployment/deploymentprocessor.go [hunk 1]
```diff
@@ -244,8 +245,9 @@ func (dp *deploymentProcessor) getApplicationAndEnvironmentForResourceID(ctx con
 // Deploy deploys the given radius resource by ordering the output resources in deployment dependency order, deploying each
 // output resource, updating static values for connections, and transforming the radius resource with computed values. It
 // returns a DeploymentOutput and an error if one occurs.
-func (dp *deploymentProcessor) Deploy(ctx context.Context, id resources.ID, rendererOutput renderers.RendererOutput) (rpv1.DeploymentOutput, error) {
+func (dp *deploymentProcessor) Deploy(ctx context.Context, id resources.ID, rendererOutput renderers.RendererOutput, operationStatus *v1.AsyncOperationStatus) (rpv1.DeploymentOutput, error) {
 	logger := ucplog.FromContextOrDiscard(ctx)
+	fmt.Println("@@@@@ Inside Deploy, operationStatus: ", operationStatus)
```
Comment (vinayada1) [inline pkg/corerp/backend/deployment/deploymentprocessor.go:250]: Ignore @@@@ for now. I will fix tests too.

**Comment by rynowak** [comment]
> > Hi @vinayada1 - what's the plan for this PR? There's no description and no associated work-item.
> 
> Added this.

Thanks 👍

**Pull request closed without merging**